[documentation] #2181: Review README (#2190)
hyperledger/iroha: README.md

# Hyperledger Iroha

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
![Rust](https://github.com/hyperledger/iroha/workflows/Rust/badge.svg?branch=iroha2-dev)
[![codecov](https://codecov.io/gh/hyperledger/iroha/branch/iroha2-dev/graph/badge.svg)](https://codecov.io/gh/hyperledger/iroha)

Iroha is a simple and efficient blockchain ledger based on the  distributed ledger technology (DLT). Its design principles are inspired by the Japanese Kaizen principle of eliminating excesses (muri).

> Iroha can help you manage your accounts, assets, on-chain data storage with efficient smart contracts, while being Byzantine- and crash-fault tolerant and being able to withstand a fault rate of 33%.

Reviewer: ```suggestion
Iroha can help you manage your accounts, assets, on-chain data storage with efficient smart contracts, while being Byzantine- and crash-fault tolerant with a fault rate of 33%.
```

Reviewer: In retrospect we can't explain this without also making it sound like Iroha crashes at a rate of 33% (which is mostly unacceptable).

## Features

Iroha is a fully-featured blockchain ledger. With Iroha you can:

* Create and manage custom fungible assets, such as currencies, gold, and others
* Create and manage non-fungible assets
* Manage user accounts with a domain hierarchy and multi-signature transactions
* Use efficient portable smart contracts implemented either via WebAssembly or Iroha Special Instructions
* Use both permissioned and permission-less blockchain deployments

> Iroha also offer you:

Reviewer: ```suggestion
Iroha offers
```

* Byzantine fault-tolerance with up to 34% fault rate  
* Efficient in-memory operations
* Extensive telemetry support out of the box

> * Modular structure that can be easily customized with `--no-default-features`

Reviewer: ```suggestion
* Modular structure. 
```

* Event-driven architecture with strongly-typed events

# System Requirements

RAM and storage requirements depend on your use case: whether you need to build or deploy a network, how big it is, and so on. This table summarises the requirements:

| Use case          | CPU               | RAM   | Storage[^1] |
|-------------------|-------------------|-------|-------------|

> | Build (minimum)   | Dual-core CPU     | 4GB   | 20GB        |

Reviewer: We will clarify these in May.

| Build (recommend) | AMD Ryzen™ 5 1600 | 16GB  | 40GB        |
| Deploy (small)    | Dual-core CPU     | 8GB+  | 20GB+       |
| Deploy (large)    | AMD Epyc™ 64-core | 128GB | 128GB+      |

> [^1]: Note that all operations are done in RAM, so in theory Iroha can work without persistent storage. However, since synchronising blocks may take a long time, we recommend adding a hard drive.

Reviewer: ```suggestion
[^1]: Note that all operations are done in RAM, so in theory Iroha can work without persistent storage. However, since synchronising blocks over the network after a power failure may take a long time, we recommend adding a hard drive.
```

> Consider the following about RAM requirements:

Reviewer: ```suggestion
Regarding RAM requirements:
```

* On average, you need 5 KiB of RAM per account. A network with 1 000 000 accounts uses 5GiB of memory.
* Each transfer or Mint instruction requires 1 KiB per instruction.
* RAM usage grows linearly, as all transactions are stored in memory. You should expect to consume more RAM with a higher TPS and uptime.

CPU considerations:

* Rust compilation highly favours multi-core CPUs such as Apple M1™, AMD Ryzen™/Threadripper™/Epyc™, and Intel Alder Lake™.
* On systems with restricted memory and many CPU cores, Iroha compilation may sometimes fail with `SIGKILL`. To avoid it, restrict the number of CPU cores using `cargo build -j <number>`, where `<number>` (without the angle brackets) is half of your RAM capacity rounded down.

# Building, testing and running Iroha

## Prerequisites

* [Rust](https://www.rust-lang.org/learn/get-started)
* (Optional) [Docker](https://docs.docker.com/get-docker/)
* (Optional) [Docker Compose](https://docs.docker.com/compose/install/)

> ## (Optional) Test the Setup

Reviewer: ```suggestion
## (Optional) Test the code
```

Reviewer: We are not testing any setup at this point. I'd also add a comment

```
Just to be safe, you should verify that your rust toolchain produces a compliant Iroha implementation. The easiest way to make sure is to test the code using the included unit and integration tests.

<details> <summary>Test Cargo and run API functional tests</summary>

> Test Cargo:

Reviewer: ```suggestion
Run included code tests.
```

# Hyperledger Iroha

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
![Rust](https://github.com/hyperledger/iroha/workflows/Rust/badge.svg?branch=iroha2-dev)
[![codecov](https://codecov.io/gh/hyperledger/iroha/branch/iroha2-dev/graph/badge.svg)](https://codecov.io/gh/hyperledger/iroha)

Iroha is a simple and efficient blockchain ledger based on the **distributed ledger technology (DLT)**. Its design principles are inspired by the Japanese Kaizen principle of eliminating excesses (*muri*).

> Iroha can help you manage your accounts, assets, on-chain data storage with efficient smart contracts, while being Byzantine- and crash-fault tolerant with a fault rate of **33%**.

Reviewer: ```suggestion
Iroha can help you manage your accounts, assets, on-chain data storage with efficient smart contracts, while being Byzantine- and crash-fault tolerant.
```

## Features

Iroha is a fully-featured blockchain ledger. With Iroha you can:

* Create and manage custom fungible assets, such as currencies, gold, and others
* Create and manage non-fungible assets
* Manage user accounts with a domain hierarchy and multi-signature transactions
* Use efficient portable smart contracts implemented either via WebAssembly or Iroha Special Instructions
* Use both permissioned and permission-less blockchain deployments

Iroha offers:

> * Byzantine fault-tolerance with up to 34% fault rate  

Author: @appetrosyan not sure why the numbers here and on line 9 are different

Reviewer: They're not the actual number is `1/3` which some people rounded up, and some-down. I think we should omit the figure on line 9 and replace this with `33%`, because `1/3` is only accurate for large numbers of peers.

bash ./scripts/test_env.sh cleanup
```

</details>

## Build Iroha

> - Build Iroha and its CLI client:

Reviewer: ```suggestion
- Build Iroha and accompanying binaries:
```

bash ./scripts/test_env.sh cleanup
```

</details>

## Build Iroha

- Build Iroha and accompanying binaries:

  ```bash
  cargo build
  ```

- (Optional) Build the latest Iroha image:

  ```bash
  docker build . -t hyperledger/iroha2:dev
  ```

>   Skipping this step will pull the Docker container from the Hyperledger DockerHub.

Reviewer: ```suggestion
If you skip this step, the Iroha container will be built using the latest available image. 
```

* [`iroha_futures`](futures) is used for `async` programming.
* [`iroha_logger`](logger) uses `tracing` to provide logging facilities.
* [`iroha_macro`](macro) provides the convenience macros.
* [`iroha_p2p`](p2p) defines peer creation and handshake logic.
* [`iroha_permissions_validators`](permissions_validators) defines permission validation logic.
* [`iroha_substrate`](substrate) is the bridge substrate `XClaim` external module.
* [`iroha_telemetry`](telemetry) is used for monitoring and analysis of telemetry data.
* [`iroha_version`](version) provides message versioning for non-simultaneous system updates.

# Maintenance

A brief overview on how to configure and maintain an Iroha instance:

- [Configuration](#configuration)
- [Endpoints](#endpoints)
- [Logging](#logging)
- [Monitoring](#monitoring)
- [Storage](#storage)
- [Scalability](#scalability)

## Configuration

You can provide configuration parameters either as a `config.json` or using environment variables. Refer to the [detailed list](./docs/source/references/config.md) of all available configuration parameters.

Configuration examples you may use as a reference point: 
- [`core/src/samples.rs`](./core/src/samples.rs)

> - [`client/src/samples.rs`](./core/src/samples.rs)

Author: @appetrosyan these links are broken, where should I refer to instead?

Reviewer: Replace both with `cli/src/samples.rs`

## Logging

By default, Iroha provides logs in a human-readable format and prints them out to `stdout`.

The logging level can be changed either via a [configuration option](./docs/source/references/config.md#loggermax_log_level) or at run-time using the `configuration` endpoint.

<details><summary>Example: changing log level</summary>

For example, if your Iroha instance is running at `127.0.0.1:8080` and you want to change the log level to `DEBUG` using `curl`, you should send a `POST` request with a JSON containing the new log level. Like this:
```bash
curl -X POST \
    -H 'content-type: application/json' \
    http://127.0.0.1:8080/configuration \
    -d '{"LogLevel": "DEBUG"}' -i
```
</details>

### JSON Logging Mode

Additionally, Iroha supports a JSON logging mode.

To enable it, provide the [logging file](./docs/source/references/config.md#loggerlog_file_path) to store the logs in. On UNIX, you can also specify `/dev/stdout` or `/dev/stderr` if you prefer to pipe the output to [`bunyan`](https://www.npmjs.com/package/bunyan).

[Log rotation](https://www.commandlinux.com/man-page/man5/logrotate.conf.5.html) is the responsibility of the peer administrator.

## Monitoring

> The details of the `Health` endpoint can be found [in API specifications](./docs/source/references/api_spec.md#health).

Reviewer: ```suggestion
The details of the `Health` endpoint can be found [in the API specifications](./docs/source/references/api_spec.md#health).
```
